libass/ass_rasterizer.c

```diff
@@ -694,72 +752,92 @@ static inline void rasterizer_fill_halfplane(const BitmapEngine *engine,
 
-    size_t n = rst->size[index] - offs;
-    struct segment *line = rst->linebuf[index] + offs;
-    if (!n) {
-        rasterizer_fill_solid(engine, buf, width, height, stride, winding);
+    size_t offs = rst->size[index] - n_lines[0] - n_lines[1];
+    struct segment *line = rst->linebuf[index] + offs, *line1 = line + n_lines[0];
+    int flags0 = get_fill_flags(line,  n_lines[0], winding[0]);
+    int flags1 = get_fill_flags(line1, n_lines[1], winding[1]);
+    int flags = (flags0 | flags1) ^ FLAG_COMPLEX;
+    if (flags & (FLAG_SOLID | FLAG_COMPLEX)) {
+        rasterizer_fill_solid(engine, buf, width, height, stride, flags & FLAG_SOLID);
+        rst->size[index] = offs;
         return true;
     }
-    if (n == 1) {
-        static const int test = SEGFLAG_UL_DR | SEGFLAG_EXACT_LEFT;
-        if (((line->flags & test) != test) == !(line->flags & SEGFLAG_DN))
-            winding++;
-
-        int flag = 0;
-        if (winding)
-            flag ^= 1;
-        if (winding - 1)
-            flag ^= 3;
-        if (flag & 1)
-            rasterizer_fill_halfplane(engine, buf, width, height, stride,
-                                      line->a, line->b, line->c,
-                                      flag & 2 ? -line->scale : line->scale);
-        else
-            rasterizer_fill_solid(engine, buf, width, height, stride, flag & 2);
+    if (!(flags & FLAG_GENERIC) && ((flags0 ^ flags1) & FLAG_COMPLEX)) {
+        if (flags1 & FLAG_COMPLEX)
+            line = line1;
+        rasterizer_fill_halfplane(engine, buf, width, height, stride,
+                                  line->a, line->b, line->c,
+                                  flags & FLAG_REVERSE ? -line->scale : line->scale);
         rst->size[index] = offs;
         return true;
     }
     if (width == 1 << engine->tile_order && height == 1 << engine->tile_order) {
-        engine->fill_generic(buf, stride, line, rst->size[index] - offs, winding);
+        if (!(flags1 & FLAG_COMPLEX)) {
+            engine->fill_generic(buf, stride, line, n_lines[0], winding[0]);
+            rst->size[index] = offs;
+            return true;
+        }
+        if (!(flags0 & FLAG_COMPLEX)) {
+            engine->fill_generic(buf, stride, line1, n_lines[1], winding[1]);
+            rst->size[index] = offs;
+            return true;
+        }
+        if (flags0 & FLAG_GENERIC)
+            engine->fill_generic(buf, stride, line, n_lines[0], winding[0]);
+        else
+            engine->fill_halfplane(buf, stride, line->a, line->b, line->c,
+                                   flags0 & FLAG_REVERSE ? -line->scale : line->scale);
+        if (flags1 & FLAG_GENERIC)
+            engine->fill_generic(rst->tile, width, line1, n_lines[1], winding[1]);
+        else
+            engine->fill_halfplane(rst->tile, width, line1->a, line1->b, line1->c,
+                                   flags1 & FLAG_REVERSE ? -line1->scale : line1->scale);
+        // XXX: should be max instead of add
```

Comment:
That lead to really miniscule difference that exist only at rare edge cases.
And correct implementation require new assembly function so it's not cost-effective.
Also I'm planning to rewrite that part for tiled-rendering anyway.